\documentclass{article}

\usepackage{listings}
\usepackage{hyperref}
\usepackage{todonotes}
\usepackage{verbatim}
\usepackage{tikz}
\usepackage{tkz-graph}
\usepackage{pgf}
\usetikzlibrary{arrows,automata}

% \newcommand*\beautier{\verb;beautier;}

\title{beautier: BEAUti for R}
\author{Rich\`el J.C. Bilderbeek, Rampal S. Etienne}


% \newcommand*\filename[1]{${#1}$}
% \newcommand*\myfilename[1]{\verb;{#1};}

% Style of code
% From http://r.789695.n4.nabble.com/How-to-nicely-display-R-code-with-the-LaTeX-package-listings-tp4648110.html
\usepackage{fancyvrb} 
\definecolor{codegreen}{rgb}{0,0.6,0}
\definecolor{codegray}{rgb}{0.5,0.5,0.5}
\definecolor{codepurple}{rgb}{0.58,0,0.82}
\definecolor{backcolor}{rgb}{0.95,0.95,0.92}
\lstdefinestyle{mystyle}{
  language=R,% set programming language
  basicstyle=\ttfamily\small,% basic font style
  commentstyle=\color{gray},% comment style
  numbers=left,% display line numbers on the left side
  numberstyle=\scriptsize,% use small line numbers
  numbersep=10pt,% space between line numbers and code
  tabsize=2,% sizes of tabs
  showstringspaces=false,% do not replace spaces in strings by a certain character
  captionpos=b,% positioning of the caption below
  breaklines=true,% automatic line breaking
  escapeinside={(*}{*)},% escaping to LaTeX
  fancyvrb=true,% verbatim code is typset by listings
  extendedchars=false,% prohibit extended chars (chars of codes 128--255)
  literate={"}{{\texttt{"}}}1{<-}{{$\bm\leftarrow$}}1{<<-}{{$\bm\twoheadleftarrow$}}1
  {~}{{$\bm\sim$}}1{<=}{{$\bm\le$}}1{>=}{{$\bm\ge$}}1{!=}{{$\bm\neq$}}1{^}{{$^{\bm\wedge}$}}1,% item to replace, text, length of chars
  alsoletter={.<-},% becomes a letter
  alsoother={$},% becomes other
  otherkeywords={!=, ~, $, \&, \%/\%, \%*\%, \%\%, <-, <<-, /},% other keywords
  deletekeywords={c}% remove keywords 
}
\lstset{style=mystyle}

\begin{document}

\maketitle

\begin{abstract}
  \textbf{1. }
  Here, we present a package, \verb;beautier;, 'BEAUti for R', for the R programming language. \\
  \textbf{2. }
    \verb;beautier; allows for scripted use of the BEAST2 phylogenetics tool, 
    by creating BEAST2 input files from an R function call. \\
  \textbf{3. }
    We describe \verb;beautier; usage, the novel functionality it provides
    compared to BEAUti, and give some minimal examples. \\
  \textbf{4. }
    As \verb;beautier; is free, libre, open-source and designed to be extended, 
    We conclude by describing the current development of the package \\
\end{abstract}

% Key-words: computational biology, evolution, phylogeny, BEAST2, R

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Phylogenies are a commonly used tool to explore evolutionary hypotheses.
Not only can phylogenies tell us how species relate to each other, also relevant parameters like
extinction and speciation rates can be estimated from them.

BEAST2 \cite{bouckaert2014beast} is a Bayesian phylogenetics tool.
BEAST2 creates a posterior of jointly-estimated phylonies and model
parameters, from a DNA, RNA or amino acid alignment. BEAST2 is a 
console application, that needs a configuration file containing alignments and model parameters.

BEAST2 is bundled with the BEAUti \cite{drummond2012bayesian}. 
BEAUti is a program to create a BEAST2 XML configuration file, 
using a user-friendly graphical user interface, with helpful and reasonable 
default settings. BEAUti replaces the manual editing of the BEAST2 XML
files. 

BEAUti cannot be called from a command-line script, which is not a problem in all cases.
For example, the BEAST book \cite{drummond2015bayesian} encourages to first infer a posterior
from simpler models first, then exploring if adding complexity changes the inferred results.
This can easily be done manually using BEAUti. 
For bigger theoretical explorations (for example, using thousands of simulated alignments), this approach
becomes inviable. 

Here, we present \verb;beautier;, 'BEAUti for R', which 
creates BEAST2 configuration files from an R function call. 
The interface of \verb;beautier; mimics BEAUti and is easy to use. This
familiar interface helps both beginner and experienced BEAST2 users 
to create configuration files from shell scripts.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{figure}
  \centering
  \begin{tikzpicture}[->,>=stealth',shorten >=1pt,auto,node distance=6cm, semithick]   
  \tikzstyle{every state}=[]
  \node[state] (A) [rectangle] {
    \includegraphics[width=0.2\textwidth]{alignment.png}
    \includegraphics[width=0.2\textwidth]{thought_cloud.png}
  };   
  \node[state] (E) [below of=A, rectangle] {
    \includegraphics[height=0.1\textheight]{xml.png}
  };   
  \node[state] (F) [rectangle, below of=E] {
    \includegraphics[width=0.3\textwidth]{DensiTreeExample2.jpg}
    \includegraphics[width=0.6\textwidth]{ParameterEstimates.png}
  };
  \path (A) edge [anchor = east] node {\includegraphics[height=0.15\textheight]{beautier_logo.png}} (E)
        (A) edge [anchor = west] node {\includegraphics[height=0.15\textheight]{BeautiSiteModel.png}} (E)
        (E) edge [anchor = east] node {\includegraphics[width=0.25\textwidth]{lumier_logo_big.png}} (F) 
        (E) edge [anchor = west] node {\includegraphics[width=0.2\textwidth]{beast_logo.png}} (F); 
  \end{tikzpicture}

  \caption{
    Workflow. From an alignment and priors, one creates a BEAST2 XML input file. This
    can be done using beautier or BEAUti. The created configuration file is run by lumier or BEAST2
    to create a posterior of phylogenies and model parameter estimates.
  }
  \label{fig:workflow}
\end{figure}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Description}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\verb;beautier; is written in the R programming language \cite{R}.
\verb;beautier; creates the BEAST2 input files from an R function call,
in a similar way that BEAUti does.

\verb;beautier;'s main function is \verb;create_beast2_input_file;, which creates
an BEAST2 input file. \verb;create_beast2_input_file; needs at least the name of a FASTA file containing a DNA alignment
and a name for the to-be-created output file. This interface follows BEAUti's default settings.
Per alignment, a site model, clock model and tree priors can be chosen.
Multiple alignments can be used, each with its own (unlinked) site model, clock model and tree prior.

\begin{table}[]
\centering
\begin{tabular}{ | l | l | }
\hline
\textbf{Name} & \textbf{Description} \\
\hline
\verb;create_beast2_input_file; & Creates a BEAST2 input file \\
\hline
\verb;create_gtr_site_model; & Create a GTR site model \cite{tavare1986some} \\
\verb;create_hky_site_model; & Create an HKY site model \cite{hasegawa1985dating} \\
\verb;create_jc69_site_model; & Create a Jukes-Cantor site model \cite{cantor1969mammalian} \\
\verb;create_tn93_site_model; & Create a TN93 site model \cite{tamura1993estimation} \\
\hline
\verb;create_rln_clock_model; & Create a relaxed log-normal clock model \cite{drummond2006relaxed} \\
\verb;create_strict_clock_model; & Create a strict clock model \cite{sarich1967immunological} \\
\hline
\verb;create_bd_tree_prior; & Create a birth-death tree prior \cite{kendall1948generalized} \\
\verb;create_cbs_tree_prior; & Create a coalescent Bayesian skyline tree prior \\
\verb;create_ccp_tree_prior; & Create a coalescent constant-population tree prior \\
\verb;create_cep_tree_prior; & Create a coalescent exponential-population tree prior \\
\verb;create_yule_tree_prior; & Create a Yule tree prior \cite{yule} \\
\hline
\verb;create_beta_distr; & Create a beta distribution \\
\verb;create_exp_distr; & Create an exponential distribution \\
\verb;create_gamma_distr; & Create a gamma distribution \\
\verb;create_inv_gamma_distr; & Create an inverse gamma distribution \\
\verb;create_laplace_distr; & Create a Laplace distribution \\
\verb;create_log_normal_distr; & Create a log-normal distribution \\
\verb;create_normal_distr; & Create a normal distribution \\
\verb;create_one_div_x_distr; & Create a 1/X distribution \\
\verb;create_poisson_distr; & Create a Poisson distribution \\
\verb;create_uniform_distr; & Create a uniform distribution \\
\hline
\end{tabular}
\caption{beautier's functions}
\label{tab:functions}
\end{table}

In total, \verb;beautier; has 59 exported functions to create
a BEAST2 configuration file. \verb;beautier; is an alternative for a majority of BEAUti use cases.
\verb;beautier; does not support the full functionality of BEAUti. Considering
BEAUti's flexibility and number of plugins, this would be a Herculean effort.
To compensate for this, an extensible software architecture is used.
\verb;beautier;s future extensions can be found on its GitHub.

BEAUti assumes that a phylogeny has a crown age that needs to be jointly-estimated
with the phylogeny and other parameters. BEAUti does not allow for fixing
a phylogeny's crown age. Before \verb;beautier;, one needs to manually edit the BEAST2 
XML configuration file, which is prone to errors. \verb;beautier;, allows easy
fixing of phylogenies' crown ages. For theoretical work, a fixed and known crown age can result
in a cleaner analysis. 

\verb;beastier; and \verb;lumier; are related packages, used in testing. 
\verb;lumier; calls BEAST2 from within R. \verb;lumier; is used to confirm that the XML files 
created by \verb;beautier; are valid. Additionally, \verb;lumier; is used to run BEAST2
to create posteriors. Using \verb;beastier;, these posteriors are checked to have an
estimated or fixed crown age.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Examples}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

In R, a package's function need to be loaded in the global namespace first:

\begin{lstlisting}[language=R, caption=Loading, label=lst:loading_beautier, floatplacement=H]
library(beautier)
\end{lstlisting}

BEAUti, and likewise \verb;beautier;, need at least a FASTA filename
and an XML output filename. In BEAUti, this is achieved by loading a FASTA file (resulting
in figure \ref{fig:simplest_beauti_usage}), then saving an output file using a common
save file dialog. In \verb;beautier;, the same is achieved by listing \ref{lst:simplest_example}:

\begin{lstlisting}[language=R, caption=Simplest example, label=lst:simplest_example, floatplacement=H]
library(beautier)
create_beast2_input_file(
  "alignment.fas",
  "beast2.xml"
)
\end{lstlisting}

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{all_default.png}
  \caption{Simplest BEAUti usage}
  \label{fig:simplest_beauti_usage}
\end{figure}

This code will create a BEAST2 file with name '\verb;beast2.xml;',
using a FASTA file with name \verb;alignment.fas;, using the same default settings as BEAUti.
The default settings are, among others, to use a Jukes-Cantor site model \cite{cantor1969mammalian}, 
a strict clock, and a Yule birth tree prior \cite{yule}. 

An example of using a different site model, clock model and tree prior is shown by listing \ref{lst:all_different}:

\begin{lstlisting}[language=R, caption=Example with different site model and clock model and tree prior, label=lst:all_different, floatplacement=H]
library(beautier)
create_beast2_input_file(
  "alignment.fas",
  "beast2.xml",
  site_models = create_hky_site_model(),
  clock_models = create_rln_clock_model(),
  tree_priors = create_bd_tree_prior()
)
\end{lstlisting}

This code uses an HKY site model, a relaxed log-normal clock model and a birth-death tree prior \cite{kendall1948generalized}.
Table \ref{tab:functions} shows an overview of all functions to create site models, clock models and tree priors.

The argument names \verb;site_models;, \verb;clock_models; and \verb;tree_priors; are plural, as each of these
can be (a list of) one or more elements. Each of these arguments must have the same number of elements, so that each alignment has its
own site model, clock model and tree prior. 

\verb;beautier; creates site models, clock models and tree priors with the same default distributions as BEAUti.
For example, a Yule tree prior assumes that birth rate likelihoods follow a uniform distribution, from minus infinity
to infinity. This assumption entails that negative and positive birth rates are just as likely, where a negative birth rate is biologically impossible. 
One may prefer to have an exponential distribution instead, as this would state that birth rates are always positive, and
higher values are less likely than lower values. To do so \verb;beautier; is shown by listing \ref{lst:diff_distr}:

\begin{lstlisting}[language=R, caption=Example with Yule tree prior with different birth rate distribution, label=lst:diff_distr, floatplacement=H]
library(beautier)
create_beast2_input_file(
  "alignment.fas",
  "beast2.xml",
  tree_priors = create_yule_tree_prior(
    birth_rate_distr = create_exp_distr()    
  )
)
\end{lstlisting}

Novel about \verb;beautier; is that it allows for specifying a fixed crown age. 
By default, a phylogeny's crown age is jointly-estimated with the other parameters.
Setting a fixed crown age is not yet possible in BEAUti directly, but it is documented how to 
manually edit the XML file to allow for a fixed crown age. 
Listing \ref{lst:fixed_crown_age} shows how to specify a fixed crown age with \verb;beautier;:

\begin{lstlisting}[language=R, caption=Example with fixed crown age, label=lst:fixed_crown_age, floatplacement=H]
create_beast2_input_file(
  "alignment.fas",
  "beast2.xml"
  posterior_crown_age = 15
)
\end{lstlisting}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{beautier development and other resources}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\verb;beautier; is free, libre and open source software available from the official R package archive at 
\url{http://cran.r-project.org/src/contrib/PACKAGES.html\#beautier}.  
\verb;beautier; is licensed under the GNU General Public License.

\verb;beautier;'s development takes place on GitHub \cite{github}, 
which is a good practice for computational scientists \cite{perez2016ten} 
and improves transparency \cite{gorgolewski2016practical}.

\verb;beautier;'s uses the Travis CI \cite{travis} 
continuous integration service, which is known to significantly 
increase the the number of bugs exposed \cite{vasilescu2015}.
\verb;beautier; has a 100\% code coverage, which correlates with code quality \cite{horgan1994,del1995correlation}. 
\verb;beautier; follows Hadley Wickham's style guide \cite{style_guide}, which improves software quality \cite{fang2001}.

\verb;beautier; is dependent on multiple packages, which are 
\verb;APE; \cite{APE}, 
\verb;beastier; \cite{beastier},
\verb;devtools; \cite{devtools},
\verb;geiger; \cite{GEIGER},
\verb;ggplot2; \cite{ggplot2},
\verb;knitr; \cite{knitr},
\verb;lumier; \cite{lumier},
\verb;phangorn; \cite{phangorn},
\verb;rmarkdown; \cite{rmarkdown},
\verb;seqinr; \cite{seqinr},
\verb;stringr; \cite{stringr},
\verb;testit; \cite{testit} and 
\verb;TreeSim; \cite{TreeSim}.

\verb;beautier;'s documentation is extensive, yet concise. All functions are documented
in the package's internal documentation. For quick use, each exported function shows a minimal example. 
For easy exploration, each exported function's documentation links to related functions.
Additionally, \verb;beautier; has a vignette that demonstrates in a longer form how
to use it. The integrity of this documentation is tested by Travis CI.
The GitHub documentation helps to get started, with a dozen examples 
of a BEAUti screenshot and the equivalent \verb;beautier; code.

\verb;beautier;'s GitHub facilitates feature requests and has guidelines how to do so.
Thanks to Travis CI, newly submitted code is expected to be accepted quicker \cite{vasilescu2015}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Citation of beautier}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Scientists using \verb;beautier; in a published paper should cite this
article. Users can additionally cite the \verb;beautier; package 
directly. Citation information can be obtained by typing:

\begin{lstlisting}[language=R]
> citation("beautier")
\end{lstlisting}

from within R.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\bibliographystyle{plain}
\bibliography{article}

\begin{thebibliography}{}

\end{thebibliography}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\end{document}
\grid
